File: 3819-count-covered-buildings/count-covered-buildings.cpp
class Solution {
public:
    int countCoveredBuildings(int n, vector<vector<int>>& arr) {
        vector<int> minXforY(n+1,-1);
        vector<int> maxXforY(n+1,-1);
        vector<int> minYforX(n+1,-1);
        vector<int> maxYforX(n+1,-1);
        for( auto& v:arr){
            int r=v[0];
            int c=v[1];
            minXforY[c]=(minXforY[c]==-1) ? r:min(r,minXforY[c]);
            maxXforY[c]=(minXforY[c]==-1) ? r:max(r,maxXforY[c]);
            minYforX[r]=(minYforX[r]==-1) ? c:min(c,minYforX[r]);
            maxYforX[r]=(maxYforX[r]==-1) ? c:max(c,maxYforX[r]);
        }
        int ans=0;
        for(auto& v:arr){
            int r=v[0];
            int c=v[1];
            if(minXforY[c]<r && maxXforY[c]>r && minYforX[r]<c  && maxYforX[r]>c ){
                ans++;
            }
        }
        return ans;
    }
};

File: 3819-count-covered-buildings/count-covered-buildings.java
class Solution {
    public int countCoveredBuildings(int n, int[][] arr) {

        int[] minXforY = new int[n + 1];
        int[] maxXforY = new int[n + 1];
        int[] minYforX = new int[n + 1];
        int[] maxYforX = new int[n + 1];

        // initialize arrays with -1
        for (int i = 0; i <= n; i++) {
            minXforY[i] = -1;
            maxXforY[i] = -1;
            minYforX[i] = -1;
            maxYforX[i] = -1;
        }

        for (int[] v : arr) {
            int r = v[0];
            int c = v[1];

            minXforY[c] = (minXforY[c] == -1) ? r : Math.min(r, minXforY[c]);
            maxXforY[c] = (maxXforY[c] == -1) ? r : Math.max(r, maxXforY[c]);

            minYforX[r] = (minYforX[r] == -1) ? c : Math.min(c, minYforX[r]);
            maxYforX[r] = (maxYforX[r] == -1) ? c : Math.max(c, maxYforX[r]);
        }

        int ans = 0;

        for (int[] v : arr) {
            int r = v[0];
            int c = v[1];

            if (minXforY[c] < r &&
                maxXforY[c] > r &&
                minYforX[r] < c &&
                maxYforX[r] > c) {

                ans++;
            }
        }

        return ans;
    }
}
